proposals/yield-in-try-catch.md

```diff
@@ -334,52 +362,135 @@ bool MoveNext()
+
+The exception object will be lifted into a field just as any other local would be and accesses to it in the generated `catch` and `when` methods will use the field. In the case the language allowed `when` clauses the pattern variables would be lifted into fields as well.
+
+For example consider this code sample:
+
+```csharp
+static IEnumerable<int> Iterator()
 {
     try
     {
         try
         {
-            yield return 1;
+            yield return M(1);
         }
         finally
         {
-            throw new Exception();
+            Console.WriteLine("Finally Inner");
         }
     }
-    catch 
+    catch (InvalidOperationException ex1) when (ex.Message.Contains("hello"))
+    {
+        Console.WriteLine("Catch1");
+    }
+    catch (Exception ex2) 
     {
-        yield return 2;
+        Console.WriteLine("Catch2");
+    }
+    finally
+    {
+        Console.WriteLine("Finally Outer");
     }
 }
+static T M<T>(T t) => t;
 ```
 
-This code would cause the statement `yield return 2` to be executed in the `Dispose` method. The state machine is not executing at this point hence it cannot be returned. Ignoring the statement would certainly be surprising the users.
+Will generate the following `Dispose` method:
 
-For these reasons this proposal will not support `yield` inside a `catch` block that is observable from a `finally`.
+```csharp
+[DebuggerHidden]
+void IDisposable.Dispose()
+{
+    int num = <>1__state;
+    if ((uint)(num - -4) > 1u && num != 1)
+    {
+        return;
+    }
+    try
+    {
+        if (num != -4 && num != 1)
+        {
+            return;
+        }
+        try
+        {
+        }
+        finally
+        {
+            <>m__Finally2();
+        }
+    }
+    catch (InvalidOperationException ex) when (<>1__ex1 = ex, <>m__When1())
+    {
+        <>m__Catch1();
+    }
+    catch (Exception ex)
+    {
+        <>1__ex2 = ex;
+        <>m__Catch1();
+    }
+    finally
+    {
+        <>m__Finally1();
+    }
+}
+```
 
-### yield inside finally
+The `<>1__ex1 = ex` in the `when` clause is not legal but the IL generated for the `when` will conceptually have this behavior.
 
-The `yield` statement inside a `finally` creates the same type of code generation issues as [catch with finally][catch-finally]. For that reason it was excluded from this proposal.
+This would add a bit of complexity the feature and it only produces observable differences when a `finally` block throws an exception on the `Dispose` path. That is likely a rare case. Further it potentially increases the complexity for developers. The `Dispose` method at first glance is likely unintuitive in that it only mirrors `finally` blocks and no other statements but that is also a very simple rule to learn. Preserving `catch` and `finally` and discussing how the code flows between them is potentially more complex.
```

Comment:
```suggestion
This would add a bit of complexity to the feature and it only produces observable differences when a `finally` block throws an exception on the `Dispose` path. That is likely a rare case. Further it potentially increases the complexity for developers. The `Dispose` method at first glance is likely unintuitive in that it only mirrors `finally` blocks and no other statements but that is also a very simple rule to learn. Preserving `catch` and `finally` and discussing how the code flows between them is potentially more complex.
```